app/src/main/java/ca/marshallasch/veil/controllers/RightMeshController.java

```diff
@@ -0,0 +1,290 @@
+        final MeshManager.DataReceivedEvent event = (MeshManager.DataReceivedEvent) e;
+
+        Sync.Message message;
+        try {
+            message = Sync.Message.parseFrom(event.data);
+        }
+        catch (InvalidProtocolBufferException e1) {
+            e1.printStackTrace();
+            return;
+        }
+
+        Sync.SyncMessageType type = message.getType();
+
+        if (type == Sync.SyncMessageType.HASH_DATA) {
+
+            Log.d("DATA_RECEIVE", message.getData().toString());
+            dataStore.syncData(message.getData());
+        } else if (type == Sync.SyncMessageType.MAPPING_MESSAGE) {
+            Log.d("DATA_RECEIVE_MAP", message.getMapping().toString());
+
+            dataStore.syncDatabase(message.getMapping());
+
+            // notify anyone interested that the data store has been updated.
+            Intent intent = new Intent(NEW_DATA_BROADCAST);
+            LocalBroadcastManager.getInstance(serviceContext).sendBroadcast(intent);
+        } else if (type == Sync.SyncMessageType.REQUEST_DATA) {
+
+            Log.d("DATA_REQUEST", "recived request for data");
+            // if someone sent a message asking for data send a responce with everything
+
+            Sync.HashData hashData = dataStore.getDataStore();
+            Sync.MappingMessage mappingMessage = dataStore.getDatabase();
+
+            // send messages to the peer.
+            Sync.Message toSend = Sync.Message.newBuilder()
+                    .setType(Sync.SyncMessageType.HASH_DATA)
+                    .setData(hashData)
+                    .build();
+
+            try {
+                meshManager.sendDataReliable(event.peerUuid, DATA_PORT, toSend.toByteArray());
+
+                toSend = Sync.Message.newBuilder()
+                        .setType(Sync.SyncMessageType.MAPPING_MESSAGE)
+                        .setMapping(mappingMessage)
+                        .build();
+
+                meshManager.sendDataReliable(event.peerUuid, DATA_PORT, toSend.toByteArray());
+
+            }
+            catch (RightMeshException e1) {
+                e1.printStackTrace();
+            }
+        }
+    }
+
+    /**
+     * Called by the {@link MeshService} when the mesh state changes. Initializes mesh connection
+     * on first call.
+     *
+     * @param meshId our own user id on first detecting
+     * @param state state which indicates SUCCESS or an error code
+     */
+    @Override
+    public void meshStateChanged(MeshId meshId, int state) {
+        switch(state) {
+            case SUCCESS: // Begin connecting
+
+                try {
+                    meshManager.bind(DATA_PORT);
+
+                    meshManager.on(DATA_RECEIVED, this::handleDataReceived);
+                    meshManager.on(PEER_CHANGED, this::handlePeerChanged);
+
+                } catch (RightMeshException e) {
+                    String status = "Error initializing the library" + e.toString();
```

Comment:
This should log the message and the string that gets displayed should be from a string resource with a place holder

Reply:
Add placeholder string + printed stack trace of exception.